components-rs/bytes.rs

```diff
@@ -0,0 +1,301 @@
+pub unsafe extern "C" fn ddog_set_span_resource_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, resource);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_set_span_type_(ptr: *mut SpanBytes, slice: *mut ZendString) {
+    set_string_field!(ptr, slice, r#type);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_meta_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(
+        ptr,
+        convert_zend_to_bytes_string(key),
+        convert_zend_to_bytes_string(val),
+        meta
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_CharSlice_span_meta_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+    val: *mut ZendString,
+) {
+    insert_hashmap!(
+        ptr,
+        convert_char_slice_to_bytes_string(key),
+        convert_zend_to_bytes_string(val),
+        meta
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, convert_zend_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_CharSlice_span_meta_(ptr: *mut SpanBytes, key: CharSlice) {
+    remove_hashmap!(ptr, convert_char_slice_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_get_span_meta_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+) -> CharSlice<'static> {
+    if ptr.is_null() {
+        return CharSlice::empty();
+    }
+
+    let span = &mut *ptr;
+
+    let key = convert_char_slice_to_bytes_string(key);
+
+    match span.meta.get(&key) {
+        Some(value) => {
+            CharSlice::from_raw_parts(value.as_str().as_ptr().cast(), value.as_str().len())
+        }
+        None => CharSlice::empty(),
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_has_span_meta_(ptr: *mut SpanBytes, key: *mut ZendString) -> bool {
+    exists_hashmap!(ptr, convert_zend_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_has_CharSlice_span_meta_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+) -> bool {
+    exists_hashmap!(ptr, convert_char_slice_to_bytes_string(key), meta);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_span_metrics_(
+    ptr: *mut SpanBytes,
+    key: *mut ZendString,
+    val: f64,
+) {
+    insert_hashmap!(ptr, convert_zend_to_bytes_string(key), val, metrics);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_add_CharSlice_span_metrics_(
+    ptr: *mut SpanBytes,
+    key: CharSlice,
+    val: f64,
+) {
+    insert_hashmap!(ptr, convert_char_slice_to_bytes_string(key), val, metrics);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_span_metrics_(ptr: *mut SpanBytes, key: *mut ZendString) {
+    remove_hashmap!(ptr, convert_zend_to_bytes_string(key), metrics);
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn ddog_del_CharSlice_span_metrics_(ptr: *mut SpanBytes, key: CharSlice) {
+    remove_hashmap!(ptr, convert_char_slice_to_bytes_string(key), metrics);
+}
```

Comment:
but that one we definitely don't need - it's covered by ddog_del_span_metrics from data-pipeline for example.